rskj-core/src/main/java/org/ethereum/core/TransactionExecutor.java

```diff
@@ -566,6 +548,81 @@ private void finalization() {
         logger.trace("tx finalization done");
     }
 
+    private void localCallFinalization() {
+        if(result == null) {
+            logger.warn("this is unexpected, a transaction executor should always have a non null result");
+            return;
+        }
+
+        if(result.getException() != null) {
+            logger.warn("Local call produced an execution error: {}",
+                    executionError != null ? executionError : "unexpected");
+            return;
+        }
+
+        logger.trace("Finalize transaction gas estimation, txHash: {}, nonce:{},", tx.getHash(), toBI(tx.getNonce()));
+
+        // Should include only LogInfo's that was added during not rejected transactions
+        List<LogInfo> logsFromNonRejectedTransactions = result.logsFromNonRejectedTransactions();
+
+        TransactionExecutionSummary.Builder summaryBuilder = TransactionExecutionSummary.builderFor(tx)
+                .gasLeftover(BigInteger.valueOf(gasLeftover))
+                .logs(logsFromNonRejectedTransactions)
+                .result(result.getHReturn());
+
+        long gasRefund = refundGas();
+
+        result.setGasUsed(getGasUsed());
+
+        TransactionExecutionSummary summary = buildTransactionExecutionSummary(summaryBuilder, gasRefund);
+
+        logger.trace("Pay total refund to sender: [{}], refund val: [{}]", tx.getSender(), summary.getRefund());
+
+        // Transfer fees to miner
+        this.paidFees = summary.getFee();
+
+        logger.trace("Processing result for gas estimation");
+
+        logs = logsFromNonRejectedTransactions;
+
+        logger.trace("tx listener for gas estimation done");
+
+        logger.trace("tx finalization for gas estimation done");
+    }
+
+    private TransactionExecutionSummary buildTransactionExecutionSummary(TransactionExecutionSummary.Builder summaryBuilder, long gasRefund) {
+        summaryBuilder
+                .gasUsed(toBI(result.getGasUsed()))
+                .gasRefund(toBI(gasRefund))
+                .deletedAccounts(result.getDeleteAccounts())
+                .internalTransactions(result.getInternalTransactions());
+
+        if (result.getException() != null) {
+            summaryBuilder.markAsFailed();
+        }
+
+        logger.trace("Building transaction execution summary");
+
+        return summaryBuilder.build();
+    }
+
+    private long refundGas() {
```

Comment:
How is this function equivalent to what was previously done on `finalization`? In particular, the old finalization did not do `addDeductedRefund`  of the `gasRefund`...  🤔 probably missing something here.

Reply:
the `deductedRefund` it's just used on `gasExactimation`, as I see it shouldn't impact on consensus, I've just extracted this part to avoid duplication, 

this is the old implementation (`mEndGas` was renamed since it isn't descriptive at all):
```java
// Accumulate refunds for suicides
result.addFutureRefund(GasCost.multiply(result.getDeleteAccounts().size(), GasCost.SUICIDE_REFUND));
long gasRefund = Math.min(result.getFutureRefund(), result.getGasUsed() / 2);
mEndGas = activations.isActive(ConsensusRule.RSKIP136) ?
                GasCost.add(mEndGas, gasRefund) :
                mEndGas + gasRefund;
```